File: packages/python/tests/_conformance.py
"""Shared helpers for the Python↔JS conformance suite (`test_conformance.py`).

Python is the reference implementation and the sole writer of the goldens
under ``conformance/golden/``. This module turns a resolved ``Diagram`` into
the canonical, language-neutral JSON form (and a BPMN export digest) that both
implementations compare against. See ``conformance/README.md`` for the contract.

The JS mirror is ``packages/js/tests/_conformance.mjs`` — keep the two in sync.
"""
from __future__ import annotations

import json
from pathlib import Path

from kymo.alignment import resolve_alignments
from kymo.bpmn_layout import layout as layout_bpmn
from kymo.dsl import parse as parse_dsl
from kymo.from_bpmn import parse as parse_bpmn
from kymo.layout import layout as apply_grid_layout
from kymo.model import Diagram
from kymo.to_bpmn import export as export_bpmn

# The canonical model serializer is the `.kymo.json` body (single source of truth);
# `model_dict` includes `layout_trees`, so the cross-language model comparison now
# covers them too. See `kymo.to_kymojson` / `docs/formats/kymo.json.md`.
from kymo.to_kymojson import model_dict as canonical_model

# conformance/ lives at the repo root, three levels above packages/python/tests.
_REPO_ROOT = Path(__file__).resolve().parents[3]
_TESTS_DIR = Path(__file__).resolve().parent
SAMPLES_DIR = _REPO_ROOT / "samples"
CORPUS_DIR = _REPO_ROOT / "conformance" / "corpus"
GOLDEN_DIR = _REPO_ROOT / "conformance" / "golden"
# .bpmn import corpus: repo samples + minimal fixtures + the vendored MIWG corpus.
FIXTURES_BPMN_DIR = _TESTS_DIR / "fixtures" / "bpmn"
CORPUS_BPMN_DIR = _TESTS_DIR / "corpus_bpmn"
EXPORT_BPMN_DIR = GOLDEN_DIR / "export_bpmn"  # committed Python-export XML (interop set)
# Stems whose Python↔JS BPMN import still diverges — skipped by the JS suite,
# tracked (with a reason) to burn down. Hand-maintained, not generated.
KNOWN_DIVERGENCES_PATH = _REPO_ROOT / "conformance" / "known_divergences.json"


def load_known_divergences() -> dict:
    """{stem: reason} of tracked, not-yet-reconciled BPMN import divergences."""
    if KNOWN_DIVERGENCES_PATH.is_file():
        return json.loads(KNOWN_DIVERGENCES_PATH.read_text(encoding="utf-8"))
    return {}


# ── Corpus discovery ────────────────────────────────────────────────────────
def corpus_files() -> list[Path]:
    """All `.kymo` corpus files: repo `samples/` + curated `conformance/corpus/`,
    de-duplicated by filename stem, sorted by stem."""
    by_stem: dict[str, Path] = {}
    for directory in (SAMPLES_DIR, CORPUS_DIR):
        if directory.is_dir():
            for path in sorted(directory.glob("*.kymo")):
                by_stem[path.stem] = path
    return [by_stem[stem] for stem in sorted(by_stem)]


def bpmn_corpus_files() -> list[Path]:
    """All `.bpmn` import-corpus files: repo `samples/`, the minimal
    `tests/fixtures/bpmn/`, and the vendored MIWG `tests/corpus_bpmn/`,
    de-duplicated by filename stem, sorted by stem."""
    by_stem: dict[str, Path] = {}
    for directory in (SAMPLES_DIR, FIXTURES_BPMN_DIR, CORPUS_BPMN_DIR):
        if directory.is_dir():
            for path in sorted(directory.glob("*.bpmn")):
                by_stem[path.stem] = path
    return [by_stem[stem] for stem in sorted(by_stem)]


def resolve(path: Path) -> Diagram:
    """Run the full front-end pipeline, mirroring ``cli.main`` /
    JS ``parseDiagram`` exactly: parse → [bpmn layout] → [grid layout] →
    [alignment resolution]."""
    diagram, layout_spec, external = parse_dsl(path.read_text(encoding="utf-8"))
    had_bpmn = bool(getattr(diagram, "bpmn_blocks", None))
    if had_bpmn:
        layout_bpmn(diagram)
    if layout_spec:
        apply_grid_layout(diagram, layout_spec, external)
    if not had_bpmn:
        resolve_alignments(diagram)
    return diagram


# ── Canonical model ──────────────────────────────────────────────────────────
def _norm(value):
    """JSON-neutral normalisation: tuples→lists, integral floats→ints
    (``5.0``→``5``), ``-0``→``0``; genuine fractions are kept so a real
    divergence surfaces instead of being rounded away."""
    if isinstance(value, bool):
        return value
    if isinstance(value, int):
        return value
    if isinstance(value, float):
        return int(value) if value.is_integer() else value
    if isinstance(value, (list, tuple)):
        return [_norm(v) for v in value]
    return value  # str | None


# `canonical_model` is imported from `kymo.to_kymojson` (the `.kymo.json` body) — the
# single source of truth, with its own field-completeness guardrail. `_norm` is kept
# here only for the BPMN digest below.


# ── BPMN import (.bpmn → model) ───────────────────────────────────────────────
def is_bpmn(diagram: Diagram) -> bool:
    return any(c.shape.startswith("bpmn-") for c in diagram.components)


def import_model(xml_text: str) -> dict:
    """Canonical model of a `.bpmn` import (already fully resolved, no layout
    pass). An importer that raises is recorded as ``{"status": "error"}`` — no
    type/message (those are language-specific), so both-error counts as a match
    while error-vs-success is a tracked divergence."""
    try:
        return canonical_model(parse_bpmn(xml_text))
    except Exception:  # noqa: BLE001 — any importer failure → uniform marker
        return {"status": "error"}


# ── BPMN export digest (model → .bpmn → re-import) ─────────────────────────────
def _edge_key(edge: dict) -> str:
    points = json.dumps(edge["points"], separators=(",", ":"))
    return "\x00".join([edge["src"], edge["dst"], edge["bpmn_flow"] or "", points])


def _digest(diagram: Diagram) -> dict:
    """Sorted, BPMN-relevant subset of a (re-imported) diagram. Sorted by id so
    re-import order — not guaranteed stable cross-language — never matters."""
    comps = sorted(
        ({"id": c.id, "shape": c.shape, "icon": c.icon,
          "size": _norm(c.size), "pos": _norm(c.pos)} for c in diagram.components),
        key=lambda c: c["id"],
    )
    regs = sorted(
        ({"id": r.id, "style": r.style, "label": r.label, "bounds": _norm(r.bounds)}
         for r in diagram.regions),
        key=lambda r: r["id"],
    )
    edges = sorted(
        ({"src": e.src, "dst": e.dst, "bpmn_flow": e.bpmn_flow, "points": _norm(e.points)}
         for e in diagram.edges),
        key=_edge_key,
    )
    return {
        "width": _norm(diagram.width),
        "height": _norm(diagram.height),
        "components": comps,
        "regions": regs,
        "edges": edges,
    }


def digest_of_xml(xml_text: str) -> dict:
    """Digest of a BPMN XML string, via the real importer. Used to compare a
    language's own export AND the committed Python export (cross-import interop)."""
    return _digest(parse_bpmn(xml_text))


def export_xml(diagram: Diagram) -> str:
    """Raw BPMN 2.0 XML from the real exporter (committed for the interop set)."""
    return export_bpmn(diagram)


def bpmn_digest(diagram: Diagram) -> dict:
    """Round-trip the diagram through the real exporter + importer, then digest.
    Exercises `to_bpmn.export`; robust to XML-formatting noise."""
    return digest_of_xml(export_xml(diagram))

File: packages/js/tests/_conformance.mjs
import "./_init.mjs";
/**
 * Shared helpers for the Python↔JS conformance suite (`conformance.test.js`).
 *
 * The JS implementation asserts its resolved model + BPMN export against the
 * committed goldens under `conformance/golden/` (written by Python, the
 * reference impl). This module is the JS mirror of
 * `packages/python/tests/_conformance.py` — keep the two in sync. It produces
 * the same canonical, language-neutral JSON shape (snake_case keys, arrays for
 * points, integral-float collapse). See `conformance/README.md`.
 */
import { existsSync, readdirSync, readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

// `canonicalModel` is the `.kymo.json` body serializer (single source of truth, incl.
// `layout_trees`); the conformance model comparison now covers layout trees too.
import { parseBpmn, toBpmn, modelDict as canonicalModel } from "../dist/index.js";
export { canonicalModel };

const here = dirname(fileURLToPath(import.meta.url));
const repoRoot = join(here, "..", "..", "..");
export const SAMPLES_DIR = join(repoRoot, "samples");
export const CORPUS_DIR = join(repoRoot, "conformance", "corpus");
export const GOLDEN_DIR = join(repoRoot, "conformance", "golden");
export const EXPORT_BPMN_DIR = join(GOLDEN_DIR, "export_bpmn");
// .bpmn import corpus: repo samples + minimal fixtures + the vendored MIWG corpus.
const FIXTURES_BPMN_DIR = join(repoRoot, "packages", "python", "tests", "fixtures", "bpmn");
const CORPUS_BPMN_DIR = join(repoRoot, "packages", "python", "tests", "corpus_bpmn");
export const KNOWN_DIVERGENCES_PATH = join(repoRoot, "conformance", "known_divergences.json");

function globByStem(dirs, ext) {
  const byStem = new Map();
  for (const dir of dirs) {
    if (!existsSync(dir)) continue;
    for (const f of readdirSync(dir).sort()) {
      if (f.endsWith(ext)) byStem.set(f.slice(0, -ext.length), join(dir, f));
    }
  }
  return [...byStem.keys()].sort().map((stem) => ({ stem, path: byStem.get(stem) }));
}

// ── Corpus discovery (mirror corpus_files / bpmn_corpus_files) ─────────────
export function corpusFiles() {
  return globByStem([SAMPLES_DIR, CORPUS_DIR], ".kymo");
}

export function bpmnCorpusFiles() {
  return globByStem([SAMPLES_DIR, FIXTURES_BPMN_DIR, CORPUS_BPMN_DIR], ".bpmn");
}

export function loadKnownDivergences() {
  return existsSync(KNOWN_DIVERGENCES_PATH)
    ? JSON.parse(readFileSync(KNOWN_DIVERGENCES_PATH, "utf8"))
    : {};
}

// ── Canonical model ────────────────────────────────────────────────────────
/** JSON-neutral normalisation: arrays recursed, `-0`→`0`; numbers otherwise
 *  kept as-is (JS has a single number type, so `5.0` is already `5`, and a
 *  genuine fraction survives to surface a real divergence). */
function norm(value) {
  if (value === null || value === undefined) return null;
  if (Array.isArray(value)) return value.map(norm);
  if (typeof value === "number") return value === 0 ? 0 : value; // collapse -0
  return value; // string | boolean
}

// `canonicalModel` (= `modelDict`) is imported above from the library — the single
// source of truth for the `.kymo.json` body. `norm` is kept here for the BPMN digest.

// ── BPMN import (.bpmn → model) ────────────────────────────────────────────
export function isBpmn(d) {
  return d.components.some((c) => String(c.shape).startsWith("bpmn-"));
}

/** Canonical model of a `.bpmn` import. Mirror of Python `import_model`: an
 *  importer that throws is recorded as `{status:"error"}` (no type/message —
 *  language-specific), so both-error matches and error-vs-success diverges. */
export function importModel(xmlText) {
  try {
    return canonicalModel(parseBpmn(xmlText));
  } catch {
    return { status: "error" };
  }
}

// ── BPMN export digest (model → .bpmn → re-import) ──────────────────────────
const byId = (a, b) => (a.id < b.id ? -1 : a.id > b.id ? 1 : 0);
const SEP = String.fromCharCode(0); // NUL separator, mirrors Python `_edge_key`'s "\x00"
const edgeKey = (e) => [e.src, e.dst, e.bpmn_flow || "", JSON.stringify(e.points)].join(SEP);

/** Sorted BPMN-relevant subset of a (re-imported) diagram — mirror of `_digest`. */
function digest(d) {
  const components = d.components
    .map((c) => ({ id: c.id, shape: c.shape, icon: c.icon, size: norm(c.size), pos: norm(c.pos) }))
    .sort(byId);
  const regions = d.regions
    .map((r) => ({ id: r.id, style: r.style, label: r.label, bounds: norm(r.bounds) }))
    .sort(byId);
  const edges = d.edges
    .map((e) => ({ src: e.src, dst: e.dst, bpmn_flow: e.bpmnFlow, points: norm(e.points) }))
    .sort((a, b) => {
      const ka = edgeKey(a), kb = edgeKey(b);
      return ka < kb ? -1 : ka > kb ? 1 : 0;
    });
  return { width: norm(d.width), height: norm(d.height), components, regions, edges };
}

/** Digest of a BPMN XML string via the real importer — used for a language's own
 *  export AND the committed Python export (cross-import interop). */
export function digestOfXml(xmlText) {
  return digest(parseBpmn(xmlText));
}

export function bpmnDigest(d) {
  return digestOfXml(toBpmn(d));
}
